src/ship_cmd.cpp

```diff
@@ -704,6 +703,8 @@ static void ShipController(Ship *v)
 
 	const uint number_of_steps = ShipAccelerate(v);
 	for (uint i = 0; i < number_of_steps; ++i) {
+		if (ShipMoveUpDownOnLock(v)) return;
```

Comment:
A ship is only considered inside a lock if it's precisely on a specific pixel within the tile. Without this extra check fast ships might skip over that pixel and the ship will just pass through the lock without waiting (and without changing z height).

Reply:
Speed limits for locks (and docks in also) is a nice idea, but out of scope of this change I think.